src/python/pants/pantsd/pailgun_server.py

```diff
@@ -178,19 +195,74 @@ def handle_request(self):
     with self.lifecycle_lock():
       self._handle_request_noblock()
 
+  def _should_poll_forever(self, timeout):
+    return timeout < 0
+
+  def _should_keep_polling(self, timeout, time_polled):
+    return self._should_poll_forever(timeout) or time_polled < timeout
+
+  @contextmanager
+  def ensure_request_is_exclusive(self, environment, request):
+    """
+    Ensure that this is the only pants running.
+
+    We currently don't allow parallel pants runs, so this function blocks a request thread until
+    there are no more requests being handled.
+    """
+    # TODO add `did_poll` to pantsd metrics
+
+    timeout = float(environment['PANTSD_REQUEST_TIMEOUT_LIMIT'])
+
+    @contextmanager
+    def yield_and_release(time_waited):
+      try:
+        self.logger.debug("request lock acquired {}.".format("on the first try" if time_waited == 0 else "in {} seconds".format(time_waited)))
+        self.free_to_handle_request_lock.clear()
+        yield
+      finally:
+        self.free_to_handle_request_lock.set()
+        self.logger.debug("released request lock.")
+
+    time_polled = 0.0
+    polling_interval = 1.0 # check the lock every second
+    self.logger.debug("request {} is trying to aquire the request lock.".format(request))
+
+    # NB: Optimistically try to acquire the lock without blocking, in case we are the only request being handled.
+    # This could be merged into the `while` loop below, but separating this special case for logging helps.
+    if self.free_to_handle_request_lock.wait(0):
+      with yield_and_release(time_polled):
+        yield
+    else:
+      # We have to wait for another request to finish being handled.
+      NailgunProtocol.send_stderr(request, "Another pants invocation is running. Will wait {} for it to finish, after which pantsd will time out.\n".format(
```

Comment:
```suggestion
      NailgunProtocol.send_stderr(request, "Another pants invocation is running. Will wait {} for it to finish before giving up.\n".format(
```

The fact that pantsd happens to be the thing giving up doesn't matter to the user here.

Reply:
Good point, thanks!